### PR TITLE
PCHR-2847:  Deletion of Public Holiday Leave Requests

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
Currently the Admin can delete other types of leave requests from the profile page of a contact except a public holiday leave request. There already exists an API to delete leave requests and this can also delete Public Holiday leave requests, except that public holiday leave requests are handled somewhat differently in L&A and they can't be deleted the same way other leave requests are deleted.
This PR allows modifies the Leave Request service in order to allow an Admin to delete Public Holiday leave requests the proper way.

## Before
- The LeaveRequestService.delete method will allow a public holiday leave request to be deleted but not in the proper way it should be deleted.

## After
- The LeaveRequestService.delete method now allows a public holiday leave request to be deleted properly.

## Technical Details
- Because when Public holiday leave requests are created, the balance change for a leave request already created on the same day as the public holiday leave request is changed to zero, so this implies that when deleting the public holiday from the system, this action must be undone. The `deleteForContact` method on the `PublicHolidayLeaveRequestDeletionService` already takes care of this, so this method was utilized in the `delete` method of the leave request service to achieve the purpose.

## Comments
- A small FE change will be needed to add the delete popup/button for Public holiday leave requests on the contact profile page.

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php [hunk 1]
```diff
@@ -233,6 +235,47 @@ public function testDeleteSoftDeletesTheLeaveRequest() {
     $this->assertEquals(1, $leaveRequestRecord->is_deleted);
   }
 
+  public function testPublicHolidayLeaveRequestIsDeletedAndBalanceRecalculatedForOverlappingLeaveRequestDate() {
+    AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php]: Do you think it might be a good idea to change these tests to use generated dates relative to the current date? It might bring up errors for edge cases when the tests are run on certain days.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php [hunk 2]
```diff
@@ -233,6 +235,47 @@ public function testDeleteSoftDeletesTheLeaveRequest() {
...
+    $this->assertEquals(-1, LeaveBalanceChange::getTotalBalanceChangeForLeaveRequest($leaveRequest));
+
+    PublicHolidayLeaveRequestFabricator::fabricate($this->leaveContact, $publicHoliday);
+    $publicHolidayLeaveRequest = LeaveRequest::findPublicHolidayLeaveRequest($this->leaveContact, $publicHoliday);
+
+    //Balance change for Leave request will be zero since a public holiday leave request is created
+    //for same date.
+    $this->assertEquals(0, LeaveBalanceChange::getTotalBalanceChangeForLeaveRequest($leaveRequest));
+
+    $this->getLeaveRequestServiceWhenCurrentUserIsAdmin()->delete($publicHolidayLeaveRequest->id);
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php]: It's probably too late to change any of this, but this test seems to be testing a lot of different services and classes. I'm guessing then that L&A uses a lot of static calls too?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php [hunk 3]
```diff
@@ -233,6 +235,47 @@ public function testDeleteSoftDeletesTheLeaveRequest() {
...
+    $publicHoliday = new PublicHoliday();
+    $publicHoliday->date = CRM_Utils_Date::processDate('2016-10-10');
+
+    $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation([
+      'from_date' => CRM_Utils_Date::processDate($publicHoliday->date),
+      'to_date' => CRM_Utils_Date::processDate($publicHoliday->date),
+      'contact_id' => $this->leaveContact,
+      'type_id' => 1,
+      'status_id' => 1
+    ], true);
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php]: Upper case for boolean

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletion.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * its dependencies
+ */
+class CRM_HRLeaveAndAbsences_Factory_PublicHolidayLeaveRequestDeletion {
+
+  /**
+   * Returns a new instance of a PublicHolidayLeaveRequestDeletion Service
+   *
+   * @return \CRM_HRLeaveAndAbsences_Service_PublicHolidayLeaveRequestDeletion
+   */
+  public static function create() {
```
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletion.php]: It shouldn't get messy.  I think this will suffice. Besides, the factory is only used in the Leave request service class.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php [hunk 5]
```diff
@@ -25,6 +27,11 @@ class CRM_HRLeaveAndAbsences_Service_LeaveRequest {
    */
   private $leaveRequestRightsService;
 
+  /**
+   * @var \CRM_HRLeaveAndAbsences_Service_PublicHolidayLeaveRequestDeletion
```
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php]: Not really, other classes are not aliased. Also aliasing it doesn't seem to work in phpstorm as I can't click the class to access it. Does it work for you?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletionTest.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * CRM_HRLeaveAndAbsences_Factory_PublicHolidayLeaveRequestDeletionTes
+ *
+ * @group headless
+ */
+class CRM_HRLeaveAndAbsences_Factory_PublicHolidayLeaveRequestDeletionTest extends BaseHeadlessTest  {
+
+  public function testCreate() {
+    $service = PublicHolidayLeaveRequestDeletionFactory::create();
```
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletionTest.php]: Well, Its a simple test to ensure that once the class changes, then this test should be updated also. We have this kind of tests for most factories in L&A.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php (hunk 1 above)
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php]: doesn't really matter, for this particular test, we are only concerned that a public holiday leave request is deleted for a particular date. 
For relative dates, we have tons of tests in the public holiday deletion service tests that already does justice to that.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php (hunk 2 above)
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php]: Yeah, isn't it everywhere in Civi :( ? But its limited in L&A though.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletion.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (mickadoo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletion.php]: Alright, in the future I think we should consider the container for this kind of thing though.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (mickadoo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php]: If you look at lines 6 - 10 of this file it's all setting aliases for classes, and line 69 which you wrote, uses the `PublicHolidayLeaveRequestDeletionFactory` alias, so I'm not sure why you don't use

`use CRM_HRLeaveAndAbsences_Service_PublicHolidayLeaveRequestDeletion as PublicHolidayLeaveRequestDeletion`

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletionTest.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (mickadoo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Factory/PublicHolidayLeaveRequestDeletionTest.php]: I don't think it adds any value, I can't think of any case where you would make a change and this test would remind the developer that it introduces an unintended bug. But if that's what you have for all factories in L&A already then ¯\\\_(ツ)\_/¯

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php [hunk 7]
```diff
@@ -59,6 +66,8 @@ public function __construct(
     $this->leaveBalanceChangeService = $leaveBalanceChangeService;
     $this->leaveRequestStatusMatrixService = $leaveRequestStatusMatrixService;
     $this->leaveRequestRightsService = $leaveRequestRightsService;
+    $publicHolidayLeaveRequestDeletionFactory = new PublicHolidayLeaveRequestDeletionFactory();
```
Comment (mickadoo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php]: I just noticed this now, but if the only method called is a static one, why are you creating a new instance here?

Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php]: Nice catch!

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php]: The aliases are used in the functions themselves and not in the function doc.
